src/libraries/Microsoft.Extensions.Configuration.CommandLine/src/CommandLineConfigurationProvider.cs

```diff
@@ -39,96 +39,159 @@ public CommandLineConfigurationProvider(IEnumerable<string> args, IDictionary<st
+        /// </remarks>
+        /// <param name="previousArg">The first string argument</param>
+        /// <param name="currentArg">The second string argument</param>
+        /// <param name="pair">
+        /// A properly resolved configuration key-value pair, or null if previous argument is invalid.
+        /// </param>
+        /// <returns>
+        /// True if the args can be resolved to a proper configuration key-value pair.
+        /// </returns>
+        private bool TryProcessArgs(string previousArg, string currentArg, out (string key, string value) pair)
+        {
+            string key, value;
+            int keyStartIndex = 0;
 
-                    // Override value when key is duplicated. So we always have the last argument win.
-                    data[key] = value;
+            if (previousArg.StartsWith("--"))
+            {
+                keyStartIndex = 2;
+            }
+            else if (previousArg.StartsWith("-"))
+            {
+                keyStartIndex = 1;
+            }
+            else if (previousArg.StartsWith("/"))
+            {
+                // "/SomeSwitch" is equivalent to "--SomeSwitch" when interpreting switch mappings
+                // So we do a conversion to simplify later processing
+                previousArg = $"--{previousArg.Substring(1)}";
+                keyStartIndex = 2;
+            }
+
+            int separator = previousArg.IndexOf('=');
+
+            if (separator < 0)
+            {
+                // If there is neither equal sign nor prefix in current argument, it is an invalid format
+                if (keyStartIndex == 0)
+                {
+                    // Ignore invalid formats
+                    pair = default;
+                    return false;
+                }
+
+                // If the switch is a key in given switch mappings, interpret it
+                if (_switchMappings != null
+                    && _switchMappings.TryGetValue(
+                        previousArg,
+                        out string mappedKey))
+                {
+                    key = mappedKey;
+                }
+                // If the switch starts with a single "-" and it isn't in given mappings,
+                // or in any other case, use the switch name directly as a key
+                else
+                {
+                    key = previousArg.Substring(keyStartIndex);
+                }
+
+                // If the argument is last in list, the next argument begins
+                // with an arg delimiter, or the next argument contains '=',
+                // then treat argument as switch and record value of "true"
+                if (currentArg == null
+                    || currentArg.StartsWith("--")
+                    || currentArg.StartsWith("-")
+                    || currentArg.StartsWith("/")
+                    || currentArg.Contains("="))
+                {
+                    value = "true";
```

Comment:
what would happen if the value expected for the key is not a boolean? Do we have validation for that?

Reply:
I'm not sure I understand the question. The `value` here gets assigned `"true"` if the next command line argument found is another key. For example:
```
./program.exe -Key1 --Key2 someValue /Key3 Key4=42
```
would generate the dictionary `{ {"Key1", "true"}, {"Key2", "someValue"}, {"Key3", "true"}, {"Key1", "42"} }`.

The absence of a switch without a value can be interpreted as a falsy value for that switch. For example, if my program is looking for a value for Key1 in `IConfiguration` to determine whether or not to enable a feature, and I removed "Key1" from the execution arguments, I would simply not enable that feature if it were missing, much like a switch on a typical *nix program.

Without a given value, `true` is the best guess. However, I could assign it pretty much any string value I want. I just picked "true" since it seemed the most intuitive. Consumers of this library can still bind the configuration to a POCO, much like appsettings.json configuration, and `{"Key1", "true"}` in the `IConfiguration` would be successfully parsed and assigned to something like `public bool Key1 { get; set; }`.